text/0000-async_await.md

```diff
@@ -0,0 +1,606 @@
+functions and calls.
+
+Async/await & futures can be a powerful abstraction for asynchronicity and
+concurrency in general, and likely has applications outside of the asynchronous
+IO space. The use cases we've experience with today are generally tied to async
+IO, but by introducing first class syntax and libstd support we believe more
+use cases for async & await will also flourish, that are not tied directly to
+asynchronous IO.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Async functions
+
+Functions can be annotated with the `async` keyword, making them "async
+functions":
+
+```rust
+async fn function(argument: &str) -> usize {
+     // ...
+}
+```
+
+Async functions work differently from normal functions. When an async function
+is called, it does not enter the body immediately. Instead, it evaluates to an
+anonymous type which implements `Future`. As that future is polled, the
+function is evaluated up to the next `await` or return point inside of it (see
+the await syntax section next).
+
+An async function is a kind of delayed computation - nothing in the body of the
+function actually runs until you begin polling the future returned by the
+function. For example:
+
+```rust
+async fn print_async() {
+     println!("Hello from print_async")
+}
+
+fn main() {
+     let future = print_async();
+     println!("Hello from main");
+     futures::block_on(future);
+}
+```
+
+This will print `"Hello from main"` before printing `"Hello from print_async"`.
+
+An `async fn foo(args..) -> T` is a function of the type
+`fn(args..) -> impl Future<Output = T>`. The return type is an anonymous type
+generated by the compiler.
+
+### `async ||` closures
+
+In addition to functions, async can also be applied to closures. Like an async
+function, an async closure has a return type of `impl Future<Output = T>`, rather
+than `T`. When you call that closure, it returns a future immediately without
+evaluating any of the body (just like an async function).
+
+```rust
+fn main() {
+    let closure = async || {
+         println("Hello from async closure.");
+    };
+    println!("Hello from main");
+    let future = closure();
+    println!("Hello from main again");
+    futures::block_on(future);
+}
+```
+
+This will print both "Hello from main" statements before printing "Hello from
+async closure."
+
+## `async` blocks
+
+You can create a future directly as an expression using an `async` block:
+
+```rust
+let my_future = async {
+    println!("Hello from an async block");
+};
+```
+
+This form is equivalent to an immediately-invoked `async` closure.
+That is:
+
+```rust
+async { /* body */ }
+
+// is equivalent to
+
+(async || { /* body */ })()
+```
+
+except that control-flow constructs like `return`, `break` and `continue` are
+not allowed within `body` (unless they appear within a fresh control-flow
+context like a closure or a loop).
```

Comment:
I think not having early return support inside an async block would be a mistake. Fundamentally an async block _is_ a form of closure (in terms of being some sort of "callable" that encloses over an environment). Which then makes an "async closure" actually a higher ordered closure that returns a closure (and "async fn" is a function that returns a closure, but because the only thing that is in scope are the arguments the closureness of the body is less relevant). I've been trying to find a more specific term for this concept, but from what I can tell there's not a term that's exactly relevant (at least from reading wikipedia, there might be something in the actual literature).

Specifically how I have imagined async blocks working is that they function as both:

* an implicit (and nameable) break point, i.e.
  ```rust
  async {
    if true { break "foo" }
  }
  ```

  ```rust
  'a: async {
      loop {
          // assuming named break with value is supported, not sure of the syntax
          break 'a "foo";
      }
  }
  ```

* a `?`-operator boundary (i.e. similar to a `do catch` block without implicit `Try::Ok` wrapping)

  ```rust
  let reader: AsyncRead = ...;
  async {
      let foo = await!(reader.read_to_end())?;
      Ok(foo.parse().unwrap_or(0))
  }: impl Future<Output = io::Result<u32>>
  ```

The second point is very important if async blocks are ever going to be used with fallible IO.

The current `async_block!` macro supports both this use cases, but uses `return` instead of `break` because of being defined as a generator (which also means no named breaking). I agree that having `return` inside an async block seems wrong, since at first glance `async {` doesn't seem like a function boundary that would stop a `return`, which is why I feel like saying it's an implicit boundary for `break` would be a better way to support early return. `continue` definitely doesn't make sense since there's no looping involved.

Reply:
But if it is a boundary for the `?`-operator, then shouldn't it use `return`? IMO the two go hand in hand.

Maybe "async block" isn't the best name for this construct, because it's not actually a control flow block. However, the construct itself makes sense, because it'll be frequently used.

Reply:
> But if it is a boundary for the `?`-operator, then shouldn't it use `return`? IMO the two go hand in hand.

Currently yes, but `do catch` blocks change that. They don't support `break`, but they do disconnect `return` expressions and the `?`-operator within their scope

```rust
fn foo() -> u32 {
    let bar: Result<&str, &str> = do catch {
        Err("fail")?;
        return 5;
        "success"
    };
    println!("{:?}", bar);
    6
}
```

> Maybe "async block" isn't the best name for this construct, because it's not actually a control flow block (it's an immediately invoked async closure).

Yep, that's why I was trying to find out if there was a more fundamental term for closure-like-things that covers things that aren't functions.

Reply:
About whether `return` and `?` go hand in hand:

> Currently yes, but do catch blocks change that.

> ? will likely very soon work with try blocks, which work at the expression level rather than the function level.

@Nemo157 @rpjohnst I agree, using `return` doesn't make sense, then.